csrc/expr_simplifier.cpp

```diff
@@ -1677,16 +1677,22 @@ Val* eliminateTrivialComputation(Val* value, const Context& context) {
       }
     }
   } else if (auto uop = dynamic_cast<UnaryOp*>(value->definition())) {
-    // -(-x) -> x, !(!x) -> x
     auto optype = uop->getUnaryOpType();
     if (optype == UnaryOpType::Neg || optype == UnaryOpType::Not) {
+      // -(-x) -> x, !(!x) -> x
       auto uop_in = dynamic_cast<UnaryOp*>(uop->in()->definition());
       if (uop_in != nullptr) {
         auto optype_in = uop_in->getUnaryOpType();
         if (optype == optype_in) {
           return uop_in->in();
         }
       }
+    } else if (optype == UnaryOpType::Abs) {
+      // abs(x) -> x if x >= 0
+      auto abs_in = uop->in();
+      if (prove::isNonNegative(abs_in, context)) {
```

Comment:
Can this also simplify `abs(x)` to `-x` where `x <= 0`?

Reply:
No, we can not do that yet. Currently, I don't see any use case for this. It is not hard to implement it, just
```C++
else if (auto neg_in = IrBuilder::negExpr(abs_in); prove::isNonNegative(neg_in, context)) {
  return neg_in;
}
```
But the `prove::isNonNegative` is not free, so I prefer not to do it if I don't see any use case. In the future, if we refactor our system to know the sign of `abs_in` in one pass regardless of if it is positive or negative, then we should add that missing simplification.